core/shared/src/main/scala/cats/effect/IO.scala

```diff
@@ -1064,6 +1064,75 @@ object IO extends IOCompanionPlatform with IOLowPriorityImplicits {
   def defer[A](thunk: => IO[A]): IO[A] =
     delay(thunk).flatten
 
+  /**
+   * Suspends an asynchronous side effect with optional immediate result in `IO`.
+   *
+   * The given function will be invoked during evaluation of the `IO` to:
+   *   - check if result is already available;
+   *   - "schedule" the asynchronous callback, where the callback of type `Either[Throwable, A]
+   *     \=> Unit` is the parameter passed to that function. Only the ''first'' invocation of
+   *     the callback will be effective! All subsequent invocations will be silently dropped.
+   *
+   * The process of registering the callback itself is suspended in `IO` (the outer `IO` of
+   * `IO[Either[Option[IO[Unit]], A]]`).
+   *
+   * The effect returns `Either[Option[IO[Unit]], A]` where:
+   *   - right side `A` is an immediate result of computation (callback invocation will be
+   *     dropped);
+   *   - left side `Option[IO[Unit]] `is an optional finalizer to be run in the event that the
+   *     fiber running `async(k)` is canceled.
+   *
+   * For example, here is a simplified version of `IO.fromCompletableFuture`:
+   *
+   * {{{
+   * def fromCompletableFuture[A](fut: IO[CompletableFuture[A]]): IO[A] = {
+   *   fut.flatMap { cf =>
+   *     IO.asyncPoll { cb =>
+   *       if (cf.isDone) {
+   *         //Register immediately available result of the completable future or handle an error
+   *         IO(cf.get)
+   *           .map(Right(_))
+   *           .handleError { e =>
+   *             cb(Left(e))
+   *             Left(None)
+   *           }
+   *       } else {
+   *         IO {
+   *           //Invoke the callback with the result of the completable future
+   *           val stage = cf.handle[Unit] {
+   *             case (a, null) => cb(Right(a))
+   *             case (_, e) => cb(Left(e))
+   *           }
+   *
+   *           //Cancel the completable future if the fiber is canceled
+   *           Left(Some(IO(stage.cancel(false)).void))
+   *         }
+   *       }
+   *     }
+   *   }
+   * }
+   * }}}
+   *
+   * @see
+   *   [[async]] for a simplified variant without an option for immediate result
+   */
+  def asyncPoll[A](
```

Comment:
The one problem I see with the name is we *already* use the word `Poll` for something quite different! This might be worth bikeshedding a bit to avoid user confusion.

Reply:
I don't think my English is good enough to provide a meaningful name, but after some thoughts there are my suggestions:

1. `asyncAttempt/tryAsync` -- as in 'we'll try to schedule async operation but if results are already available, then so be it'. Both with 'error handling' connotations, sadly.
2. `resolve/resolveEventually` -- as in 'computed value will be resolved one way or another, right now or in future';
3. `registerAsyncCallback` -- intentionally enterprise-like name to emphasis that this is a low-level method like `cont`.

Reply:
I'll give this some thought and report back in a day or two! I actually rather like `asyncAttempt`, but you're right it does have that negative "error" context.

Reply:
@djspiewak, did you had time to think about a name?

Reply:
Ah, I haven't! Thank you for the reminder. I'll get to this, I apologize.

Reply:
I think `asyncCheckAttempt` is probably the best I can come up with. I agree with you that we don't necessarily *want* it to be a really concise name.

```suggestion
  def asyncCheckAttempt[A](
```

Reply:
Done.